### ISSUE TITLE
dietpi software install stuck when installing software (pivpn)

### ISSUE THREAD

**Issue by kelliegator** [MichaIng/DietPi#3844]
#### Required Information
- DietPi version | `cat /boot/dietpi/.version`
G_DIETPI_VERSION_CORE=6
G_DIETPI_VERSION_SUB=33
G_DIETPI_VERSION_RC=3
G_GITBRANCH='master'
G_GITOWNER='MichaIng'

- Distro version | `echo $G_DISTRO_NAME` or `cat /etc/debian_version`
buster
10.6

- Kernel version | `uname -a`
Linux DietPi 5.4.51-v7+ #1333 SMP Mon Aug 10 16:45:19 BST 2020 armv7l GNU/Linux

- SBC model | `echo $G_HW_MODEL_NAME` or (EG: RPi3)
RPi 3 Model B (armv7l)

- Power supply used | (EG: 5V 1A RAVpower)
Standard 5V power supply.

- SDcard used | (EG: SanDisk ultra)
Sandisk 16GB Ultra

#### Additional Information (if applicable)
- Software title | (EG: Nextcloud)
PiVPN

- Was the software title installed freshly or updated/migrated?
Installed freshly.

- Can this issue be replicated on a fresh installation of DietPi?
I don't think so.

- Bug report ID | `echo $G_HW_UUID`

Doing a bugreport just resulted in this screen, so I don't know if I'm doing anything wrong.
<img width="556" alt="Skärmavbild 2020-10-20 kl  10 16 36" src="https://user-images.githubusercontent.com/8292200/96559490-6bb3da80-12bd-11eb-909b-01dcb1c220a8.png">

#### Steps to reproduce

1. Install PiVPN with OpenVPN.
2. Uninstall PiVPN.
3. Install it again (with the intent to install Wireguard instead).

#### Expected behaviour

The installation should go normally and let me install PiVPN again but with Wireguard instead.

#### Actual behaviour

The installation gets stuck at "./install.bash".  I've even tried to install this software and leave the computer on overnight, but when I checked the next morning it was still stuck at the same step, so I don't know what that is about.

#### Extra details
I had originally installed PiVPN with OpenVPN, but I uninstalled 'cause I wanted to reinstall PiVPN try Wireguard instead, but every time I try to install PiVPN it gets stuck. I also tried just installing Wireguard and it was very slow, but I didn't have time to wait and check if it also got stuck like with PiVPN. I'm a little worried that I've broken my Pi, or something. I'm not the most tech savvy in the world even though I've learned a lot over time.

But everything else with my Pi seems to work fine, like Nextcloud and samba, so I don't know what to believe.

I hope I've been able to provide all the necessary info. If I've missed something, feel free to point it out.

**Comment by Joulinar** [comment]
Hi,

many thanks for your message. Pls can you have a look to your system time? Usually certificate issues happen if system time is not up-to-date. But this did not explain the issue with installation of PiVPN. Pls can you start installation again and let it run for a couple of minutes. Open a 2nd SSH connection and check if there is a `whiptail` process

`ps -ef |grep whiptail`

If you have such a process, PiVPN installer is waiting on user input.

At least on my very quit testing, it seems the SSH session was hanging. I hit a couple of times "ENTER" key and out of the sudden, the `whiptail` was showing up 🤔 

****

**EDIT:**
I did some further testing and indeed the `whiptail` is not showing up. But the install process is waiting on input. Even if it is not shown, it's reacting on input. Every time hitting "ENTER" key, the `whiptail` is changing. 

```
root@DietPi3:~# ps -ef |grep whiptail
root      1625  1329  0 11:03 pts/0    00:00:00 whiptail --msgbox --backtitle Welcome --title PiVPN Automated Installer This installer will transform your Raspberry Pi into an OpenVPN or WireGuard server! 31 118
```
```
root@DietPi3:~# ps -ef |grep whiptail
root      1629  1329  1 11:03 pts/0    00:00:00 whiptail --msgbox --backtitle Initiating network interface --title Static IP Needed The PiVPN is a SERVER so it needs a STATIC IP ADDRESS to function properly.  In the next section, you can choose to use your current network settings (DHCP) or to manually edit them. 31 118
```
```
root@DietPi3:~# ps -ef |grep whiptail
root      1650  1329  0 11:04 pts/0    00:00:00 whiptail --backtitle Calibrating network interface --title DHCP Reservation --yesno --defaultno Are you Using DHCP Reservation on your Router/DHCP Server? These are your current Network Settings:  ???IP address:    192.168.0.12/24 ???Gateway:       192.168.0.1  Yes: Keep using DHCP reservation No: Setup static IP address Don't know what DHCP Reservation is? Answer No. 31 118
```
```
root@DietPi3:~# ps -ef |grep whiptail
root      1654  1329  0 11:04 pts/0    00:00:00 whiptail --backtitle Calibrating network interface --title Static IP Address --yesno Do you want to use your current network settings as a static address?  ????IP address:    192.168.0.12/24 ????Gateway:       192.168.0.1 31 118
```

Don't know why but sooner or later it's going to pop up

@MichaIng 
do you know how this could happen?

**Comment by kelliegator** [comment]
> Hi,
> 
> many thanks for your message. Pls can you have a look to your system time? Usually certificate issues happen if system time is not up-to-date. But this did not explain the issue with installation of PiVPN. Pls can you start installation again and let it run for a couple of minutes. Open a 2nd SSH connection and check if there is a `whiptail` process

Hey. I feel really stupid here but I actually don't quite know how to check the system time on my Dietpi. I'm still a bit of a newbie at this.

As for whiptail, this is the result I got:

`root     10253  9393  0 15:52 pts/0    00:00:00 whiptail --msgbox --backtitle Welcome --title PiVPN Automated Installer This installer will transform your Raspberry Pi into an OpenVPN or WireGuard server! 20 70
root     10432 10404  0 15:52 pts/1    00:00:00 grep whiptail`

**Comment by MichaIng** [comment]
This is a bug our side, we wrapped the installer into an error handler but forgot to keep printing it's output to console.
This fixes it: #3845

**Comment by Joulinar** [comment]
@kelliegator 
as workaround, you cold continue hitting "ENTER" key. This should move you through the installer once stuck. At least on my testing, it was needed to hit "ENTER" 6,7,8 times before it's going to show up. 😃 

regarding system time, just use `date` on command line to check it

**Comment by MichaIng** [comment]
Changelog: https://github.com/MichaIng/DietPi/commit/de269e92ca55e0459ef4ec92ca203beb7b58daf5
Quick fix:
```
sed -i 's|^[[:blank:]]*G_EXEC_NOEXIT=1 G_EXEC \./install\.bash|\t\t\tG_EXEC_OUTPUT=1 G_EXEC_NOEXIT=1 G_EXEC ./install.bash|' /boot/dietpi/dietpi-software
```

**Comment by Joulinar** [comment]
thx @MichaIng 
even better than hitting "ENTER" on a blank screen 🤣

**Comment by MichaIng** [comment]
MOTD patch added (and removed all the obsolete pre-v6.33 ones 😉):
```
[[ -w '/boot/dietpi/dietpi-software' ]] && grep '^[[:blank:]]*G_EXEC_NOEXIT=1 G_EXEC \./install\.bash' /boot/dietpi/dietpi-software &> /dev/null && sed -i 's|^[[:blank:]]*G_EXEC_NOEXIT=1 G_EXEC \./install\.bash|\t\t\tG_EXEC_OUTPUT=1 G_EXEC_NOEXIT=1 G_EXEC ./install.bash|' /boot/dietpi/dietpi-software &> /dev/null
```

Jep whiptail and all other installer output shows up again. Lol but I face an error:
```
:::    Checking for iptables-persistent... not installed!
:::    Failed to install iptables-persistent!
```
`iptables-persistent` install fails. No APT output, this is why I don't like this `debconf-apt-progress` command, it suppresses all the helpful output and only exists with error 🤔.

... lol no idea why it fails within the installer, running it from console works pretty fine:
```
:::    Checking for iptables-persistent... not installed!
:::    Failed to install iptables-persistent!
[FAILED] DietPi-Software | ./install.bash
2020-10-20 19:40:51 root@VM-Buster:/tmp/DietPi-Software# apt install iptables-persistent
Reading package lists... Done
Building dependency tree
Reading state information... Done
iptables-persistent is already the newest version (1.0.11).
0 upgraded, 0 newly installed, 0 to remove and 1 not upgraded.
1 not fully installed or removed.
After this operation, 0 B of additional disk space will be used.
Do you want to continue? [Y/n] y
Setting up iptables-persistent (1.0.11) ...
```
Let me know if it goes through in your case, maybe my testing VM is upset.
... ah when you faced the whiptails, it did already finished nicely, so indeed my VM is upset 😅.

**Comment by Joulinar** [comment]
yes, I have seen a couple of reports regarding `iptables-persistent` as well on PiVPN side. However it's usually working if your run the installer again.

Found this one on our forum https://dietpi.com/phpbb/viewtopic.php?f=11&t=8122&p=27591&hilit=iptables+persistent#p27591

**Comment by MichaIng** [comment]
We could debug this, running manual `iptables-persistent` install on the machine where the installed failed with this error, then step-wise applying the environment of the PiVPN installer 🤔.

**Comment by kelliegator** [comment]
> Changelog: [de269e9](https://github.com/MichaIng/DietPi/commit/de269e92ca55e0459ef4ec92ca203beb7b58daf5)
> Quick fix:
> 
> ```
> sed -i 's|^[[:blank:]]*G_EXEC_NOEXIT=1 G_EXEC \./install\.bash|\t\t\tG_EXEC_OUTPUT=1 G_EXEC_NOEXIT=1 G_EXEC ./install.bash|' /boot/dietpi/dietpi-software
> ```

This worked for me. Thanks!

**Comment by mgagnequebec** [comment]
6.33.3 - this just happened to me when selecting multiple installs (netdata, rpi-monitor, webmin). The Enter key trick got me through the first package but then I got this: 

https://imgur.com/a/Bi7mu2J

I know these screenshots are not great, but I chose the option "send log to the dev" also. 

while it was stuck, I also tried to ssh with my other pc that's on the same network/subnet  (to try the "ps -ef |grep whiptail" command) but it timed out after asking me for the ssh password (using openssh, not bear)

Not sure what else I can put in this comment other than the fact that I tried to update to 6.34.0 but it says that i'm up to date,  I hope this helps

 edit: ah, nvm. seen the fix is in the beta now
